File: src/threshold.ts
import type { Finding, Hotspot, Config } from './types.js';

export function checkThresholds(
  config: Config,
  findings: Finding[],
  hotspots: Hotspot[]
): { passed: boolean; reasons: string[] } {
  const reasons: string[] = [];

  for (const target of config.failOn) {
    switch (target) {
      case 'nullpointers': {
        const npe = findings.filter(f => f.source === 'spotbugs' && f.rule.startsWith('NP_'));
        if (npe.length > 0)
          reasons.push(`${npe.length} null pointer path${npe.length > 1 ? 's' : ''} found`);
        break;
      }
      case 'complexity': {
        const cx = findings.filter(f =>
          f.source === 'pmd' &&
          ['CyclomaticComplexity', 'CognitiveComplexity', 'NPathComplexity'].includes(f.rule)
        );
        if (cx.length > 0)
          reasons.push(`${cx.length} complexity violation${cx.length > 1 ? 's' : ''}`);
        break;
      }
      case 'dead-code': {
        const dead = findings.filter(f =>
          f.source === 'pmd' &&
          ['UnusedLocalVariable','UnusedPrivateField','UnusedPrivateMethod',
           'UnusedFormalParameter','UnnecessaryImport','UnusedImports'].includes(f.rule)
        );
        if (dead.length > 0)
          reasons.push(`${dead.length} dead code issue${dead.length > 1 ? 's' : ''}`);
        break;
      }
      case 'bugs': {
        const bugs = findings.filter(f => f.source === 'spotbugs' && f.severity !== 'low');
        if (bugs.length > 0)
          reasons.push(`${bugs.length} SpotBugs finding${bugs.length > 1 ? 's' : ''}`);
        break;
      }
      case 'hotspots': {
        if (config.maxHotspotScore !== Infinity) {
          const hot = hotspots.filter(h => h.score > config.maxHotspotScore);
          if (hot.length > 0)
            reasons.push(`${hot.length} file${hot.length > 1 ? 's' : ''} exceed hotspot score ${config.maxHotspotScore}`);
        }
        break;
      }
      case 'all': {
        const crit = findings.filter(f => f.severity === 'critical' || f.severity === 'high');
        if (crit.length > 0)
          reasons.push(`${crit.length} critical/high finding${crit.length > 1 ? 's' : ''}`);
        break;
      }
    }
  }

  return { passed: reasons.length === 0, reasons };
}

File: crates/reap/src/threshold.rs
use crate::config::Config;
use crate::types::{Severity, Source};
use crate::JavaResult;

fn plural(n: usize, noun: &str) -> String {
    format!("{n} {noun}{}", if n > 1 { "s" } else { "" })
}

pub fn check_thresholds(config: &Config, result: &JavaResult) -> (bool, Vec<String>) {
    let reasons: Vec<String> = config
        .fail_on
        .iter()
        .filter_map(|t| check_one(t, config, result))
        .collect();
    (reasons.is_empty(), reasons)
}

fn check_one(target: &str, config: &Config, result: &JavaResult) -> Option<String> {
    match target {
        "nullpointers" => {
            let n = result
                .findings
                .iter()
                .filter(|f| f.source == Source::Spotbugs && f.rule.starts_with("NP_"))
                .count();
            (n > 0).then(|| format!("{} found", plural(n, "null pointer path")))
        }
        "bugs" => {
            let n = result
                .findings
                .iter()
                .filter(|f| f.source == Source::Spotbugs && f.severity != Severity::Low)
                .count();
            (n > 0).then(|| plural(n, "SpotBugs finding"))
        }
        "hotspots" => {
            if config.max_hotspot_score.is_infinite() {
                return None;
            }
            let n = result.hotspots.iter().filter(|h| h.score > config.max_hotspot_score).count();
            (n > 0).then(|| {
                format!("{} exceed hotspot score {}", plural(n, "file"), config.max_hotspot_score)
            })
        }
        "complexity" => {
            let n = result.complexity.len();
            (n > 0).then(|| format!("{} over threshold", plural(n, "complex function")))
        }
        "large-functions" => {
            let n = result.large.len();
            (n > 0).then(|| plural(n, "large function"))
        }
        "circular" => {
            let n = result.cycles.len();
            (n > 0).then(|| plural(n, "circular dependency"))
        }
        "unused-files" => {
            let n = result.unused_files.len();
            (n > 0).then(|| plural(n, "unused file"))
        }
        "unused-exports" => {
            let n = result.unused_exports.len();
            (n > 0).then(|| plural(n, "unused export"))
        }
        "duplicates" => {
            let n = result.clone_groups.len();
            (n > 0).then(|| plural(n, "duplicate block"))
        }
        "all" => {
            let n = result
                .findings
                .iter()
                .filter(|f| matches!(f.severity, Severity::Critical | Severity::High))
                .count();
            (n > 0).then(|| plural(n, "critical/high finding"))
        }
        _ => None,
    }
}
